build.assets/tooling/cmd/build-apt-repos/apt_repo_tool.go

```diff
@@ -0,0 +1,178 @@
+package main
+
+import (
+	"io/fs"
+	"log"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+)
+
+type AptRepoTool struct {
+	config       *Config
+	aptly        *Aptly
+	s3Manager    *s3manager
+	supportedOSs *map[string][]string
+}
+
+// Instantiates a new apt repo tool instance and performs any required setup/config.
+func NewAptRepoTool(config *Config, supportedOSs *map[string][]string) (*AptRepoTool, error) {
+	art := &AptRepoTool{
+		config:       config,
+		s3Manager:    NewS3Manager(*config.bucketName),
+		supportedOSs: supportedOSs,
+	}
+
+	aptly, err := NewAptly()
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create a new aptly instance")
+	}
+	art.aptly = aptly
+
+	return art, nil
+}
+
+// Runs the tool, performing all of the business logic outside of argument parsing.
+func (art *AptRepoTool) Run() error {
+	start := time.Now()
+	logrus.Infoln("Starting APT repo build process...")
+
+	err := art.s3Manager.DownloadExistingRepo(*art.config.localBucketPath)
+	if err != nil {
+		return trace.Wrap(err, "failed to sync existing repo from S3 bucket")
+	}
+
+	repos, err := art.createRepos()
+	if err != nil {
+		return trace.Wrap(err, "failed to create repos")
+	}
+
+	err = art.importNewDebs(repos)
+	if err != nil {
+		return trace.Wrap(err, "failed to import new debs")
+	}
+
+	err = art.publishRepos(repos)
+	if err != nil {
+		return trace.Wrap(err, "failed to publish repos")
+	}
+
+	aptlyRootDir, err := art.aptly.GetRootDir()
+	if err != nil {
+		return trace.Wrap(err, "failed to get Aptly root directory")
+	}
+
+	err = art.s3Manager.UploadBuiltRepo(filepath.Join(aptlyRootDir, "public"))
+	if err != nil {
+		return trace.Wrap(err, "failed to sync changes to S3 bucket")
+	}
+
+	execution_duration := time.Since(start).Round(time.Millisecond)
+	logrus.Infof("APT repo build process completed in %s\n", execution_duration)
+	return nil
+}
+
+func (art *AptRepoTool) publishRepos(repos []*Repo) error {
+	// Build a map keyed by os info with value of all repos that support the os in the key
+	// This will be used to structure the publish command
+	categorizedRepos := make(map[string][]*Repo)
+	for _, r := range repos {
+		if osRepos, ok := categorizedRepos[r.OSInfo()]; ok {
+			categorizedRepos[r.OSInfo()] = append(osRepos, r)
+		} else {
+			categorizedRepos[r.OSInfo()] = []*Repo{r}
+		}
+	}
+
+	for osInfo, osRepoList := range categorizedRepos {
+		if len(osRepoList) < 1 {
+			continue
+		}
+
+		err := art.aptly.PublishRepos(osRepoList, osRepoList[0].os)
+		if err != nil {
+			return trace.Wrap(err, "failed to publish for os \"%s\"", osInfo)
```

Comment:
```suggestion
			return trace.Wrap(err, "failed to publish for os %q", osInfo)
```

Reply:
Wow - `%q` makes logging quoted string significantly easier to read. Updated everywhere